Explorer/Assets/DCL/Multiplayer/Movement/Systems/MultiplayerMovementMessageBus.cs

```diff
@@ -45,55 +53,47 @@ private void OnMessageReceived(ReceivedMessage<Decentraland.Kernel.Comms.Rfc4.Mo
                 if (cancellationTokenSource.Token.IsCancellationRequested)
                     return;
 
-                var message = MovementMessage(receivedMessage.Payload);
+                NetworkMovementMessage message = MovementMessage(receivedMessage.Payload);
                 Inbox(message, receivedMessage.FromWalletId);
             }
         }
 
-        public void Send(NetworkMovementMessage message)
+        private void OnMessageReceived(ReceivedMessage<MovementCompressed> receivedMessage)
         {
-            WriteAndSend(message, messagePipesHub.IslandPipe());
-            WriteAndSend(message, messagePipesHub.ScenePipe());
-        }
+            if (isDisposed)
+            {
+                ReportHub.LogError(ReportCategory.MULTIPLAYER, "Receiving a message while disposed is bad");
+                return;
+            }
 
-        public void InjectWorld(World world)
-        {
-            this.globalWorld = world;
+            using (receivedMessage)
+            {
+                if (cancellationTokenSource.Token.IsCancellationRequested)
+                    return;
+
+                CompressedNetworkMovementMessage message = new ()
+                {
+                    temporalData = receivedMessage.Payload.TemporalData,
+                    movementData = receivedMessage.Payload.MovementData,
+                };
+
+                Inbox(message.Decompress(), receivedMessage.FromWalletId);
+            }
         }
 
-        private void WriteAndSend(NetworkMovementMessage message, IMessagePipe messagePipe)
+        public void Send(NetworkMovementMessage message)
         {
-            var messageWrap = messagePipe.NewMessage<Decentraland.Kernel.Comms.Rfc4.Movement>();
-            WriteToProto(message, messageWrap.Payload);
-            messageWrap.SendAndDisposeAsync(cancellationTokenSource.Token).Forget();
+            WriteAndSend(message, messagePipesHub.IslandPipe());
+            WriteAndSend(message, messagePipesHub.ScenePipe());
         }
 
-        private static void WriteToProto(NetworkMovementMessage message, Decentraland.Kernel.Comms.Rfc4.Movement movement)
+        public void InjectWorld(World world)
```

Comment:
Could you pass the world via ctor?

Reply:
this class is created when Plugins are created and globalWorld is not created at this point. However, I can call `InjectToWorld `in the `InitializeWorldRelatedModules`  of the `DynamicWorldContainer`. It will move the call from Update just to one call at bootstrap